src/dp.rs

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Differential privacy (DP) primitives.
+use crate::field::FieldElement;
+use crate::vdaf::prg::Seed;
+use std::fmt::Debug;
+
+// TODO(tholop): We could also use pairs of integers, or even
+// floats to store the DiscreteGaussian DP parameters?
+// Using BigInts might make it harder to communicate `DiscreteGaussian` to the Collector.
+use num_rational::BigRational;
+
+/// DP-related errors.
+#[derive(Debug, thiserror::Error)]
+pub enum DpError {
+    /// Error while sampling from a specific noise distribution
+    #[error("sampling error")]
+    SamplingError,
+}
+
+/// Trait for a noise distribution over a (vector space over a) finite field.
+/// Useful for differential privacy.
+/// Inherited by specific mechanism instantiations such as `[DiscreteGaussian]`
+/// that will enforce the VDAF-dependent DP guarantees.
+pub trait Dist: Clone + Debug {
+    /// Sample `len` elements from the distribution over `F`.
+    /// The randomness comes from a Prg seeded by `seed`.
+    /// TODO(tholop): do we need a generic SEED_SIZE here, or 16 is fine?
+    fn sample<F: FieldElement, const SEED_SIZE: usize>(
+        &self,
+        seed: Seed<SEED_SIZE>,
+        len: usize,
+    ) -> Result<Vec<F>, DpError>;
+
+    // TODO(tholop): do we need another method to send the distribution parameters
+    // to the Collector?
+}
+
+/// Discrete Gaussian noise distribution with mean 0 and standard deviation `sigma`.
+/// `sigma` can be computed as a function of DP parameters `epsilon`, `delta`
+/// and `sensitivity`. `sensitivity` depends on the VDAF and potentially the number of measurements.
+/// `sigma` is sufficient to characterize the distribution, but other parameters
+/// are kept for interpretability.
+///
+/// The distribution is defined over Z, represented by BigInts
+/// To implement `Dist`, the `sample` method can project on a finite F if necessary,
+/// as long as the DP guarantees are the same as the mechanism over Z
+/// (the proof is implementation dependent, e.g. DP post-processing)
+///
+/// TODO(tholop): depending on how we handle `num_measurements`, rename `sensitivity`
+/// to something like `base_sensitivity` or `sensitivity_when_num_measurements_is_1`,
+/// and explain that `add_noise_to_agg_share` is in charge of scaling the noise properly.
+pub struct DiscreteGaussian {
+    epsilon: BigRational,
+    delta: BigRational,
+    sigma: BigRational,
+    sensitivity: BigRational,
+}
+
+impl DiscreteGaussian {
+    pub fn new(epsilon: BigRational, delta: BigRational, sensitivity: BigRational) -> Self {
```

Comment:
the DP flavour of the guarantee given in thm 4 of the [discrete gaussian paper](https://arxiv.org/pdf/2004.00010.pdf) is [zero-concentrated DP (zCDP)](https://programming-dp.com/ch8.html#zero-concentrated-differential-privacy), not epsilon-delta (ED). those can be converted into each other, but for composition (e.g. executing in a loop like for gradient descent) converting for every execution and then accumulating ED gives worse bounds than accumulating zCDP and converting to ED in the end. i hence suggest using the zCDP parameter here, confusingly also called epsilon in the paper. alternatively, i suppose an implementation of a [slightly involved](https://github.com/IBM/discrete-gaussian-differential-privacy/blob/master/cdp2adp.py) method to compute a tighter zCDP parameter given epsilon and delta could be provided, based on section 2.2 of the paper.

Reply:
Yes, as noted in the comment on the latest commit, I agree that a zCDP constructor would be good, potentially in addition to an Approximate DP/ED constructor (because some people might not need composition, or prefer a simpler definition of DP).

In that case, I think that the `DiscreteGaussian` struct should hold only `sigma`, and no other metadata? We can always provide some convenience functions to convert `sigma` from/to different DP definitions.

Reply:
My only feedback here is that we want the API to be flexible enough to support multiple mechanisms. I'm not sure what's needed, so I'll just trust y'all's judgement here.